pylint/checkers/base/function_checker.py

```diff
@@ -0,0 +1,222 @@
+
+        # check if function has a contextmanager decorator
+        contextmanager_str = "contextlib.contextmanager"
+        should_be_added_to_contextmanagers = any(
+            decorator
+            # this is fully qualified import name
+            for decorator in node.decoratornames()
+            if decorator == contextmanager_str
+        )
+        # cant be a contextmanager and a erroring generator at the same time
+        if (
+            should_be_added_to_contextmanagers
+            and self._node_fails_contextmanager_cleanup(node, yield_nodes=yield_nodes)
+        ):
+            self._add_node_to_contextmanagers(node)
+            return
+
+        # otherwise could be a generator that uses a contextmanager
+        with_nodes = list(node.nodes_of_class(nodes.With))
+        async_with_nodes = list(node.nodes_of_class(nodes.AsyncWith))
+        if not with_nodes and not async_with_nodes:
+            # no inner contextmanager, so no need to check
+            return
+        all_nodes = chain(with_nodes, async_with_nodes)
+        for with_node in all_nodes:
+            if self._node_fails_generator_checks(with_node):
+                self._add_node_to_generators(with_node, node)
+
+    @staticmethod
+    def _node_fails_contextmanager_cleanup(
+        node: nodes.FunctionDef, yield_nodes: list[nodes.Yield]
+    ) -> bool:
+        """Check if a node fails contextmanager cleanup.
+
+        Current checks for a contextmanager:
+            - only if the context manager yields something other than None
+            - only if the context manager lacks a finally, or does not catch GeneratorExit
+
+        :param node: Node to check
+        :type node: nodes.FunctionDef
+        :return: True if fails, False otherwise
+        :param yield_nodes: List of Yield nodes in the function body
+        :type yield_nodes: list[nodes.Yield]
+        :rtype: bool
+        """
+        # if context manager yields a non-constant value, then continue checking
+        if any(
+            yield_node.value is None or isinstance(yield_node.value, nodes.Const)
+            for yield_node in yield_nodes
+        ):
+            return False
+        # if function body has multiple Try, filter down to the ones that have a yield node
+        try_with_yield_nodes = [
+            try_node
+            for try_node in node.nodes_of_class(nodes.Try)
+            if list(try_node.nodes_of_class(nodes.Yield))
+        ]
+        if not try_with_yield_nodes:
+            # no try blocks at all, so checks after this line do not apply
+            return True
+        # if the contextmanager has a finally block, then it is fine
+        if all(try_node.finalbody for try_node in try_with_yield_nodes):
+            return False
+        # if the contextmanager catches GeneratorExit, then it is fine
+        if all(
+            any(
+                isinstance(handler.type, nodes.Name)
+                and handler.type.name == "GeneratorExit"
```

Comment:
Check me on this, but I think elsewhere we check the `qname()` against `builtins.GeneratorExit`, something like that, just to be extra careful in case someone has polluted their namespace 😆